// @jpmorganchase/onyx-ssi-sdk/lib/cjs/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyTypeError = exports.DIDMethodFailureError = exports.JsonParseError = exports.SchemaValidationFailureError = exports.ReadFileJsonFailureError = exports.AxiosRequestFailureError = void 0;
class AxiosRequestFailureError extends Error {
    constructor(message) {
        super(message);
        this.name = 'AxiosRequestFailureError';
        Error.captureStackTrace(this, AxiosRequestFailureError);
    }
}
exports.AxiosRequestFailureError = AxiosRequestFailureError;
class ReadFileJsonFailureError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ReadFileJsonFailureError';
        Error.captureStackTrace(this, ReadFileJsonFailureError);
    }
}
exports.ReadFileJsonFailureError = ReadFileJsonFailureError;
class SchemaValidationFailureError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SchemaValidationFailureError';
        Error.captureStackTrace(this, SchemaValidationFailureError);
    }
}
exports.SchemaValidationFailureError = SchemaValidationFailureError;
class JsonParseError extends Error {
    constructor(message) {
        super(message);
        this.name = 'JsonParseError';
        Error.captureStackTrace(this, JsonParseError);
    }
}
exports.JsonParseError = JsonParseError;
class DIDMethodFailureError extends Error {
    constructor(message) {
        super(message);
        this.name = 'DIDMethodFailureError';
        Error.captureStackTrace(this, DIDMethodFailureError);
    }
}
exports.DIDMethodFailureError = DIDMethodFailureError;
class KeyTypeError extends Error {
    constructor(message) {
        super(message);
        this.name = 'KeyTypeError';
        Error.captureStackTrace(this, KeyTypeError);
    }
}
exports.KeyTypeError = KeyTypeError;
